File: api/lua_grpc/pinnacle/output.lua
---The protobuf absolute path prefix
local prefix = "pinnacle.output." .. require("pinnacle").version .. "."
local service = prefix .. "OutputService"

---@type table<string, { request_type: string?, response_type: string? }>
---@enum (key) OutputServiceMethod
local rpc_types = {
    SetLocation = {},
    ConnectForAll = {
        response_type = "ConnectForAllResponse",
    },
    Get = {
        response_type = "GetResponse",
    },
    GetProperties = {
        response_type = "GetPropertiesResponse",
    },
}

---Build GrpcRequestParams
---@param method OutputServiceMethod
---@param data table
---@return GrpcRequestParams
local function build_grpc_request_params(method, data)
    local req_type = rpc_types[method].request_type
    local resp_type = rpc_types[method].response_type

    ---@type GrpcRequestParams
    return {
        service = service,
        method = method,
        request_type = req_type and prefix .. req_type or prefix .. method .. "Request",
        response_type = resp_type and prefix .. resp_type,
        data = data,
    }
end

---@class OutputHandleModule
local output_handle = {}

---@class OutputHandle
---@field private config_client Client
---@field name string The unique name of this output
local OutputHandle = {}

---@class OutputModule
---@field private handle OutputHandleModule
local output = {}
output.handle = output_handle

---@class Output
---@field private config_client Client
local Output = {}

---Get all outputs.
---
---@return OutputHandle[]
function Output:get_all()
    local response = self.config_client:unary_request(build_grpc_request_params("Get", {}))

    ---@type OutputHandle[]
    local handles = {}

    for _, output_name in ipairs(response.output_names or {}) do
        table.insert(handles, output_handle.new(self.config_client, output_name))
    end

    return handles
end

---@param name string The name of the port the output is connected to
---@return OutputHandle | nil
function Output:get_by_name(name)
    local handles = self:get_all()

    for _, handle in ipairs(handles) do
        if handle.name == name then
            return handle
        end
    end

    return nil
end

---@return OutputHandle | nil
function Output:get_focused()
    local handles = self:get_all()

    for _, handle in ipairs(handles) do
        if handle:props().focused then
            return handle
        end
    end

    return nil
end

---@param callback fun(output: OutputHandle)
function Output:connect_for_all(callback)
    local handles = self:get_all()
    for _, handle in ipairs(handles) do
        callback(handle)
    end

    self.config_client:server_streaming_request(build_grpc_request_params("ConnectForAll", {}), function(response)
        local output_name = response.output_name
        local handle = output_handle.new(self.config_client, output_name)
        callback(handle)
    end)
end

---@param loc { x: integer?, y: integer? }
function OutputHandle:set_location(loc)
    self.config_client:unary_request(build_grpc_request_params("SetLocation", {
        output_name = self.name,
        x = loc.x,
        y = loc.y,
    }))
end

---@alias Alignment
---| "top_align_left"
---| "top_align_center"
---| "top_align_right"
---| "bottom_align_left"
---| "bottom_align_center"
---| "bottom_align_right"
---| "left_align_top"
---| "left_align_center"
---| "left_align_bottom"
---| "right_align_top"
---| "right_align_center"
---| "right_align_bottom"

---@param other OutputHandle
---@param alignment Alignment
function OutputHandle:set_loc_adj_to(other, alignment)
    local self_props = self:props()
    local other_props = other:props()

    if not self_props.x or not other_props.x then
        -- TODO: notify
        return
    end

    local alignment_parts = {}

    for str in alignment:gmatch("%a+") do
        table.insert(alignment_parts, str)
    end

    ---@type "top"|"bottom"|"left"|"right"
    local dir = alignment_parts[1]
    ---@type "top"|"bottom"|"center"|"left"|"right"
    local align = alignment_parts[3]

    ---@type integer
    local x
    ---@type integer
    local y

    if dir == "top" or dir == "bottom" then
        if dir == "top" then
            y = other_props.y - self_props.pixel_height
        else
            y = other_props.y + other_props.pixel_height
        end

        if align == "left" then
            x = other_props.x
        elseif align == "center" then
            x = other_props.x + math.floor((other_props.pixel_width - self_props.pixel_width) / 2)
        elseif align == "bottom" then
            x = other_props.x + (other_props.pixel_width - self_props.pixel_width)
        end
    else
        if dir == "left" then
            x = other_props.x - self_props.pixel_width
        else
            x = other_props.x + other_props.pixel_width
        end

        if align == "top" then
            y = other_props.y
        elseif align == "center" then
            y = other_props.y + math.floor((other_props.pixel_height - self_props.pixel_height) / 2)
        elseif align == "bottom" then
            y = other_props.y + (other_props.pixel_height - self_props.pixel_height)
        end
    end

    self:set_location({ x = x, y = y })
end

---@class OutputProperties
---@field make string?
---@field model string?
---@field x integer?
---@field y integer?
---@field pixel_width integer?
---@field pixel_height integer?
---@field refresh_rate integer?
---@field physical_width integer?
---@field physical_height integer?
---@field focused boolean?
---@field tags TagHandle[]

---Get all properties of this output.
---@return OutputProperties
function OutputHandle:props()
    local response =
        self.config_client:unary_request(build_grpc_request_params("GetProperties", { output_name = self.name }))

    local handles = require("pinnacle.tag").handle.new_from_table(self.config_client, response.tag_ids)

    response.tags = handles
    response.tag_ids = nil

    return response
end

---@return Output
function output.new(config_client)
    ---@type Output
    local self = {
        config_client = config_client,
    }
    setmetatable(self, { __index = Output })
    return self
end

---Create a new `OutputHandle` from its raw name.
---@param output_name string
function output_handle.new(config_client, output_name)
    ---@type OutputHandle
    local self = {
        config_client = config_client,
        name = output_name,
    }
    setmetatable(self, { __index = OutputHandle })
    return self
end

return output

File: api/rust_grpc/src/output.rs
use futures::{
    channel::mpsc::UnboundedSender, executor::block_on, future::BoxFuture, FutureExt, StreamExt,
};
use pinnacle_api_defs::pinnacle::{
    output::{
        self,
        v0alpha1::{
            output_service_client::OutputServiceClient, ConnectForAllRequest, SetLocationRequest,
        },
    },
    tag::v0alpha1::tag_service_client::TagServiceClient,
};
use tonic::transport::Channel;

use crate::tag::TagHandle;

#[derive(Debug, Clone)]
pub struct Output {
    channel: Channel,
    fut_sender: UnboundedSender<BoxFuture<'static, ()>>,
}

impl Output {
    pub fn new(channel: Channel, fut_sender: UnboundedSender<BoxFuture<'static, ()>>) -> Self {
        Self {
            channel,
            fut_sender,
        }
    }

    fn create_output_client(&self) -> OutputServiceClient<Channel> {
        OutputServiceClient::new(self.channel.clone())
    }

    fn create_tag_client(&self) -> TagServiceClient<Channel> {
        TagServiceClient::new(self.channel.clone())
    }

    pub fn get_all(&self) -> impl Iterator<Item = OutputHandle> {
        let mut client = self.create_output_client();
        let tag_client = self.create_tag_client();
        block_on(client.get(output::v0alpha1::GetRequest {}))
            .unwrap()
            .into_inner()
            .output_names
            .into_iter()
            .map(move |name| OutputHandle {
                client: client.clone(),
                tag_client: tag_client.clone(),
                name,
            })
    }

    pub fn get_focused(&self) -> Option<OutputHandle> {
        self.get_all()
            .find(|output| matches!(output.props().focused, Some(true)))
    }

    pub fn connect_for_all(&self, mut for_all: impl FnMut(OutputHandle) + 'static + Send) {
        for output in self.get_all() {
            for_all(output);
        }

        let mut client = self.create_output_client();
        let tag_client = self.create_tag_client();

        self.fut_sender
            .unbounded_send(
                async move {
                    let mut stream = client
                        .connect_for_all(ConnectForAllRequest {})
                        .await
                        .unwrap()
                        .into_inner();

                    while let Some(Ok(response)) = stream.next().await {
                        let Some(output_name) = response.output_name else {
                            continue;
                        };

                        let output = OutputHandle {
                            client: client.clone(),
                            tag_client: tag_client.clone(),
                            name: output_name,
                        };

                        for_all(output);
                    }
                }
                .boxed(),
            )
            .unwrap();
    }
}

#[derive(Clone, Debug)]
pub struct OutputHandle {
    pub(crate) client: OutputServiceClient<Channel>,
    pub(crate) tag_client: TagServiceClient<Channel>,
    pub(crate) name: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Alignment {
    TopAlignLeft,
    TopAlignCenter,
    TopAlignRight,
    BottomAlignLeft,
    BottomAlignCenter,
    BottomAlignRight,
    LeftAlignTop,
    LeftAlignCenter,
    LeftAlignBottom,
    RightAlignTop,
    RightAlignCenter,
    RightAlignBottom,
}

impl OutputHandle {
    pub fn set_location(&self, x: Option<i32>, y: Option<i32>) {
        let mut client = self.client.clone();
        block_on(client.set_location(SetLocationRequest {
            output_name: Some(self.name.clone()),
            x,
            y,
        }))
        .unwrap();
    }

    pub fn set_loc_adj_to(&self, other: &OutputHandle, alignment: Alignment) {
        let self_props = self.props();
        let other_props = other.props();

        let attempt_set_loc = || -> Option<()> {
            let other_x = other_props.x?;
            let other_y = other_props.y?;
            let other_width = other_props.pixel_width? as i32;
            let other_height = other_props.pixel_height? as i32;

            let self_width = self_props.pixel_width? as i32;
            let self_height = self_props.pixel_height? as i32;

            use Alignment::*;

            let x: i32;
            let y: i32;

            if let TopAlignLeft | TopAlignCenter | TopAlignRight | BottomAlignLeft
            | BottomAlignCenter | BottomAlignRight = alignment
            {
                if let TopAlignLeft | TopAlignCenter | TopAlignRight = alignment {
                    y = other_y - self_height;
                } else {
                    // bottom
                    y = other_y + other_height;
                }

                match alignment {
                    TopAlignLeft | BottomAlignLeft => x = other_x,
                    TopAlignCenter | BottomAlignCenter => {
                        x = other_x + (other_width - self_width) / 2;
                    }
                    TopAlignRight | BottomAlignRight => x = other_x + (other_width - self_width),
                    _ => unreachable!(),
                }
            } else {
                if let LeftAlignTop | LeftAlignCenter | LeftAlignBottom = alignment {
                    x = other_x - self_width;
                } else {
                    x = other_x + other_width;
                }

                match alignment {
                    LeftAlignTop | RightAlignTop => y = other_y,
                    LeftAlignCenter | RightAlignCenter => {
                        y = other_y + (other_height - self_height) / 2;
                    }
                    LeftAlignBottom | RightAlignBottom => {
                        y = other_y + (other_height - self_height);
                    }
                    _ => unreachable!(),
                }
            }

            self.set_location(Some(x), Some(y));

            Some(())
        };

        attempt_set_loc();
    }

    pub fn props(&self) -> OutputProperties {
        let mut client = self.client.clone();
        let response = block_on(
            client.get_properties(output::v0alpha1::GetPropertiesRequest {
                output_name: Some(self.name.clone()),
            }),
        )
        .unwrap()
        .into_inner();

        OutputProperties {
            make: response.make,
            model: response.model,
            x: response.x,
            y: response.y,
            pixel_width: response.pixel_width,
            pixel_height: response.pixel_height,
            refresh_rate: response.refresh_rate,
            physical_width: response.physical_width,
            physical_height: response.physical_height,
            focused: response.focused,
            tags: response
                .tag_ids
                .into_iter()
                .map(|id| TagHandle {
                    client: self.tag_client.clone(),
                    output_client: self.client.clone(),
                    id,
                })
                .collect(),
        }
    }

    // TODO: make a macro for the following or something

    pub fn make(&self) -> Option<String> {
        self.props().make
    }

    pub fn model(&self) -> Option<String> {
        self.props().model
    }

    pub fn x(&self) -> Option<i32> {
        self.props().x
    }

    pub fn y(&self) -> Option<i32> {
        self.props().y
    }

    pub fn pixel_width(&self) -> Option<u32> {
        self.props().pixel_width
    }

    pub fn pixel_height(&self) -> Option<u32> {
        self.props().pixel_height
    }

    pub fn refresh_rate(&self) -> Option<u32> {
        self.props().refresh_rate
    }

    pub fn physical_width(&self) -> Option<u32> {
        self.props().physical_width
    }

    pub fn physical_height(&self) -> Option<u32> {
        self.props().physical_height
    }

    pub fn focused(&self) -> Option<bool> {
        self.props().focused
    }

    pub fn tags(&self) -> Vec<TagHandle> {
        self.props().tags
    }
}

#[derive(Clone, Debug)]
pub struct OutputProperties {
    pub make: Option<String>,
    pub model: Option<String>,
    pub x: Option<i32>,
    pub y: Option<i32>,
    pub pixel_width: Option<u32>,
    pub pixel_height: Option<u32>,
    pub refresh_rate: Option<u32>,
    pub physical_width: Option<u32>,
    pub physical_height: Option<u32>,
    pub focused: Option<bool>,
    pub tags: Vec<TagHandle>,
}
